ui/panels-plugin/src/plugins/scatterplot/ScatterChartPanel.tsx

```diff
@@ -0,0 +1,186 @@
+
+export interface EChartTraceValue extends Omit<TraceValue, 'startTimeUnixMs'> {
+  startTime: Date;
+}
+
+export function getUnixTimeRange(timeRange: AbsoluteTimeRange) {
+  const { start, end } = timeRange;
+  return {
+    start: Math.ceil(getUnixTime(start)),
+    end: Math.ceil(getUnixTime(end)),
+  };
+}
+
+const generateErrorAlert = (message: string) => {
+  const alertMessage = {
+    name: message,
+    message: message,
+  };
+  return (
+    <div data-testid="ScatterChartPanel_ErrorAlert">
+      <ErrorAlert error={alertMessage} />
+    </div>
+  );
+};
+
+export type ScatterChartPanelProps = PanelProps<ScatterChartOptions>;
+
+/**
+ * ScatterChartPanel receives data from the DataQueriesProvider and transforms it
+ * into a `dataset` object that Apache ECharts can consume. Additionally,
+ * data formatting is also dictated in this component. Formatting includes
+ * datapoint size and color.
+ *
+ * Documentation for data structures accepted by Apache ECharts:
+ *  https://echarts.apache.org/handbook/en/concepts/dataset
+ *
+ * Examples for scatter chart formatting in Apache ECharts:
+ *  https://echarts.apache.org/examples/en/index.html#chart-type-scatter
+ *
+ * @returns a `ScatterPlot` component that contains an EChart which will handle
+ * visuzliation of the data.
+ */
+export function ScatterChartPanel(props: ScatterChartPanelProps) {
+  const { contentDimensions, definition } = props;
+  const { absoluteTimeRange } = useTimeRange();
+  const { queryResults: traceResults, isLoading: traceIsLoading } = useDataQueries('TraceQuery');
+  const queries = definition?.spec?.queries;
+
+  // Generate dataset
+  // Transform Tempo API response to fit 'dataset' structure from Apache ECharts
+  // https://echarts.apache.org/handbook/en/concepts/dataset
+  const dataset = useMemo(() => {
+    const traceData = traceResults[0]?.data;
+    if (traceIsLoading || traceData === undefined) {
+      return [];
+    }
+    const dataset = [];
+    for (const result of traceResults) {
+      if (traceIsLoading || traceData === undefined) {
+        return [];
+      }
+      if (result.isLoading || result.data === undefined) continue;
+      const dataSeries = result.data.traces.map((trace) => {
+        const newTraceValue: EChartTraceValue = {
+          ...trace,
+          startTime: new Date(trace.startTimeUnixMs), // convert unix epoch time to Date
+        };
+        return newTraceValue;
+      });
+      dataset.push({
+        source: dataSeries,
+      });
+    }
+    return dataset;
+  }, [traceIsLoading, traceResults]);
+
+  // Error check: Tempo API only supports queries from the past 7 days (as of Dec. 17, 2023)
+  const containsTempoQuery = () => {
+    let isTempoQuery = false;
+    queries?.forEach((query) => {
+      if (query.spec.plugin.kind === 'TempoTraceQuery') {
+        isTempoQuery = true;
+      }
+    });
+    return isTempoQuery;
+  };
+  const isGreaterThanSevenDays = (): boolean => {
```

Comment:
This should probably be handled in the query plugin similar to how there is Prometheus specific logic here: https://github.com/perses/perses/blob/main/ui/prometheus-plugin/src/plugins/prometheus-time-series-query/get-time-series-data.ts

Reply:
Good catch. In the latest commit, I removed this function after discovering that the Tempo API allows time-range queries over seven days. Users can override this default configuration. I double-checked this with one of my colleagues at Red Hat, who works on the backend with Tempo.